tmt/base.py

```diff
@@ -1665,13 +1672,44 @@ def import_plan(self) -> Optional['Plan']:
                 else:
                     tmt.utils.git_clone(plan_id.url, destination, self)
                 if plan_id.ref:
-                    self.run(Command('git', 'checkout', plan_id.ref), cwd=destination)
+                    # attempt to evaluate dynamic reference
+                    try:
+                        # Would it hurt if we simply rewrite plan_id.ref? Better choose safer path.
+                        plan_id._ref_expanded = resolve_dynamic_ref(
+                            workdir=destination,
+                            ref=plan_id.ref,
+                            plan=self,
+                            logger=self._logger
+                            )
+                    except tmt.utils.FileError as error:
+                        raise tmt.utils.DiscoverError(
+                            f"Failed to resolve dynamic ref of '{plan_id.ref}'.") from error
+                    if plan_id._ref_expanded:
+                        self.run(
+                            Command(
+                                'git',
+                                'checkout',
+                                plan_id._ref_expanded),
+                            cwd=destination)
                 if plan_id.path:
                     destination = destination / plan_id.path.unrooted()
                 node = fmf.Tree(str(destination)).find(plan_id.name)
 
             # Use fmf cache for exploring plans (the whole git repo is not needed)
             else:
+                # fmf does not support dynamic ref so when working only with
+                # the fmf cache we won't be able to read and process the dynamic_ref file
+                # pointing us to the right branch.
+                # The whole point of working with cache is to speed up things so it
+                # probably doesn't make sense to clone whole repo.
+                # Therefore, we have decided to lie here use a default branch instead.
+                # An alternative would be either to enforce full clonning or to fail.
+                if plan_id.ref and plan_id.ref.startswith('@'):
```

Comment:
Does it make any sense to continue here?
I would rather fail ...

Reply:
I don't know.. Depends on what is the primary reason for doing `dry` run and if we do any harm.

Reply:
> Also this is not any kind of `dry` mode, so the warning message is a bit misleading.

@psss Well, the code states `self.my_run.opt('dry'):` so if this is not a dry mode then I would say the code itself is a bit misleading.

Reply:
> So this `else` branch is used for quick exploration of plans, that is it affects `tmt`, `tmt plan ls`, `tmt plan show`. Only `tmt plan show` displays details which are affected by dynamic reference. Shall we show the warning only there? 

I am afraid that I do not understand the `tmt` code good enough to make this decision.

Reply:
> Also, I've noticed that the dynamic `ref` is not shown in the verbose output:

This is because you are hitting the `dry` case and the dynamic reference is not evaluated, see the printed warning.

Reply:
Hm, this is getting a bit complicated. Let's see. There are three cases:

* `tmt run`
* `tmt run --dry`
* `tmt plans ls` and friends

What about showing the warning only for the last case? The text could be:

    warn: Cannot evaluate dynamic reference used in the remote
    plan '/plans/dynamic-ref'. Using the default branch instead.